# Welcome to Lab: Simpson's Paradox 🔬

Before this lab section, you learned about Simpson's Paradox and confounding variables in lecture.  This week, you will find Simpson's Paradox through analysis of a dataset in Python, and get some practice writing conditionals for pandas DataFrames! 

A few tips to remember:

- **You are not alone on your journey in learning programming!**  You have your lab Teaching Assistant, your Course Aides, your lab group, and the professors (Prof. Wade and Prof. Karle), who are all here to help you out!
- If you find yourself stuck for more than a few minutes, ask a neighbor or course staff for help!  When you are giving help to your neighbor, explain the **idea and approach** to the problem without sharing the answer itself so they can have the same **<i>ah-hah</i>** moment!
- We are here to help you!  Don't feel embarrassed or shy to ask us for help!

Let's get started!

In [2]:
# Meet your CAs and TA if you haven't already!
# ...first name is enough, we'll know who they are! :)
ta_name = ""
ca1_name = ""
ca2_name = ""
ca3_name = ""

# Say hello to each other!
# - Groups of 3 are ideal :)
# - However, groups of 2 or 4 are fine too!
#
# Question of the Day (QOTD) to Ask Your Group: "Which city do you like better: Champaign or Urbana?"
partner1_name = "Gabriella"
partner1_netid = ""
partner1_city = ""

partner2_name = "Melissa"
partner2_netid = ""
partner2_city = ""

partner3_name = ""
partner3_netid = ""
partner3_city = ""

<hr style="color: #DD3403;">

# Part 1: The GPA Dataset

Many of you have likely come across one of the GPA visualizations found at https://waf.cs.illinois.edu/discovery/gpa/, either out of 
curiosity or the need to investigate a mysterious GenEd class you signed up for. 

Regardless, these visualizations are built on the **GPA Dataset** of UIUC students across all course subjects. Today, you're going to do some analysis and discover a case of Simpson's Paradox within data taken from UIUC courses!

## Load the GPA Dataset

The most recent version of the "GPA Dataset" is available here:
```
https://waf.cs.illinois.edu/discovery/gpa.csv
```

Use Python to load this dataset into a DataFrame called `df`:

In [3]:
import pandas as pd
df = pd.read_csv("https://waf.cs.illinois.edu/discovery/gpa.csv")
df

Year    Term YearTerm Subject  Number                   Course Title  \
0      2024  Winter  2024-wi     ADV     150    Introduction to Advertising   
1      2024  Winter  2024-wi    AFRO     100   Intro to African American St   
2      2024  Winter  2024-wi    ASTR     330          Extraterrestrial Life   
3      2024  Winter  2024-wi    ATMS     120   Severe and Hazardous Weather   
4      2024  Winter  2024-wi    BADM     275          Operations Management   
...     ...     ...      ...     ...     ...                            ...   
69107  2010  Summer  2010-su    STAT     410  Statistics and Probability II   
69108  2010  Summer  2010-su    STAT     440    Statistical Data Management   
69109  2010  Summer  2010-su     TAM     212          Introductory Dynamics   
69110  2010  Summer  2010-su     TAM     251   Introductory Solid Mechanics   
69111  2010  Summer  2010-su     TAM     335   Introductory Fluid Mechanics   

      Sched Type  A+   A  A-  ...  C+  C  C-  D+  D  D-  F  W  \
0            ONL  22  18   1  ...   0  0   0   0  0   0  0  0   
1            ONL   5   8   6  ...   0  1   1   0  0   0  1  0   
2            ONL   1  16   8  ...   1  0   2   0  1   0  0  0   
3            ONL   3  16   3  ...   2  2   1   1  1   1  2  0   
4            ONL   0  49  17  ...   2  3   0   0  0   0  0  0   
...          ...  ..  ..  ..  ...  .. ..  ..  .. ..  .. .. ..   
69107        LEC   5  10   2  ...   0  1   3   0  0   0  2  1   
69108        LEC   4  12   8  ...   0  0   0   0  0   0  0  0   
69109        LEC   0   1   3  ...   5  1   1   0  2   0  1  0   
69110        LCD   1   2   2  ...   3  3   2   0  0   1  1  0   
69111        LEC   5   8   9  ...   0  4   0   0  1   0  2  0   

             Primary Instructor  Students  
0                Hall, Steven R        42  
1          McMillion, Desiree Y        25  
2                Dunne, Bryan C        37  
3             Choate, Jessica J        46  
4                     Hu, Junqi        81  
...                         ...       ...  
69107        Stepanov, Alexei G        31  
69108              Unger, David        28  
69109         Morgan, William T        28  
69110  Ott-Monsivais, Stephanie        21  
69111          Keane, Richard D        39  

[69112 rows x 23 columns]

### 🔬 Test Case Checkpoint 🔬

In [4]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(len(df) == 69112), "This is not the GPA dataset you're looking for."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Data Cleaning: An Additional Column

The GPA dataset contains raw GPA data and is not the easiest to work with if we want to analyze intricacies such as the average GPA or passing rate.  Luckily, DataFrames are modifiable, so we can **add more columns** based on what questions we want to answer.

The process of modifying a dataset via deletion (cleaning up empty/unwanted values) or addition (adding new columns) is often called **data cleaning**. This is an important concept in Data Science, because you won't always receive your data in the perfect format for your purposes.

With the GPA dataset, we want to investigate one innocent question posed by a theoretical incoming student: 
- *Is it easier to get an A in STAT or CS courses at UIUC?*

To do this, we'll need to first perform some modifications on our loaded DataFrame, `df`. Specifically, we will need to create:
- An `A_Grades` column, containing the total number of students receiving an A+, A, or A- in every course

### Puzzle 1.1: Creating an `A_Grades` Column

Create the new column `A_Grades` that stores the total number of "A"s given in every course.

- We consider an "A" to be any type of A, in other words "A+", "A", or an "A-" are all included.
- In our `df`, the number of students recieving a particular grade in a course is found in the `df['Grade']` column. For example, `df['A']` contains the number of students who recieved an A.

In [5]:
df['A_Grades'] = (df['A'] + df['A+'] + df['A-'])

### 🔬 Test Case Checkpoint 🔬

In [6]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert('A_Grades' in df), "Make sure you've named the A_Grades column properly and added it to the DataFrame."
assert(df['A_Grades'].sum() == 2382654), "Double check the values of your A_Grades column."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


# Working with the GPA Dataset
Now that we have that extra column set up, we can perform basic mathematical analysis on the GPA Dataset to get insight towards our question:
- *Is it easier to get an A in STAT or CS courses at UIUC?*

Should be simple, right? Let's just see which subject, STAT or CS, has a greater percentage of A grades.

### Puzzle 1.2: Subject DataFrames

Select only the rows of the GPA dataset `df` with a `Subject` of `STAT`. Assign these rows to a new DataFrame, `df_STAT`. 
- Make sure your result only contains STAT courses!

In [7]:
df_STAT = df[df.Subject == "STAT"]

df_STAT

Year    Term YearTerm Subject  Number                   Course Title  \
2680   2023    Fall  2023-fa    STAT     100                     Statistics   
2681   2023    Fall  2023-fa    STAT     100                     Statistics   
2682   2023    Fall  2023-fa    STAT     100                     Statistics   
2683   2023    Fall  2023-fa    STAT     107         Data Science Discovery   
2684   2023    Fall  2023-fa    STAT     107         Data Science Discovery   
...     ...     ...      ...     ...     ...                            ...   
69104  2010  Summer  2010-su    STAT     100                     Statistics   
69105  2010  Summer  2010-su    STAT     100                     Statistics   
69106  2010  Summer  2010-su    STAT     400   Statistics and Probability I   
69107  2010  Summer  2010-su    STAT     410  Statistics and Probability II   
69108  2010  Summer  2010-su    STAT     440    Statistical Data Management   

      Sched Type  A+    A   A-  ...   C  C-  D+  D  D-   F  W  \
2680         LCD  49  260  128  ...   9   2   2  0   3   8  0   
2681         LCD  44  177   77  ...  10   3   3  1   0   5  0   
2682         LCD   6   84   32  ...   5   2   2  0   0   0  0   
2683         LEC  56  285   59  ...  11   6   1  4   0  10  1   
2684         LEC  66  301   49  ...  11   6   2  2   0   3  1   
...          ...  ..  ...  ...  ...  ..  ..  .. ..  ..  .. ..   
69104        LCD   1   13    3  ...   2   2   3  1   1   1  0   
69105        LCD   0   14    3  ...   3   0   0  1   0   3  1   
69106        LEC   4   15    7  ...   2   2   0  1   0   3  1   
69107        LEC   5   10    2  ...   1   3   0  0   0   2  1   
69108        LEC   4   12    8  ...   0   0   0  0   0   0  0   

       Primary Instructor  Students  A_Grades  
2680      Rao, V.N. Vimal       608       437  
2681      Rao, V.N. Vimal       437       298  
2682         Clement, Kit       195       122  
2683    Flanagan, Karle A       524       400  
2684    Flanagan, Karle A       525       416  
...                   ...       ...       ...  
69104  Hirtz, Nathaniel R        39        17  
69105    Dalpiaz, David M        28        17  
69106      Monrad, Ditlev        47        26  
69107  Stepanov, Alexei G        31        17  
69108        Unger, David        28        24  

[1044 rows x 24 columns]

Select only the rows of the GPA dataset `df` with a `Subject` of `CS`. Assign these rows to a new DataFrame, `df_CS`. 
- Make sure your result only contains CS courses!

In [8]:
df_CS = df[df.Subject == "CS"]

df_CS

Year    Term YearTerm Subject  Number                   Course Title  \
886    2023    Fall  2023-fa      CS     100   Computer Science Orientation   
887    2023    Fall  2023-fa      CS     101   Intro Computing: Engrg & Sci   
888    2023    Fall  2023-fa      CS     105      Intro Computing: Non-Tech   
889    2023    Fall  2023-fa      CS     124    Intro to Computer Science I   
890    2023    Fall  2023-fa      CS     128   Intro to Computer Science II   
...     ...     ...      ...     ...     ...                            ...   
68988  2010  Summer  2010-su      CS     101   Intro Computing: Engrg & Sci   
68989  2010  Summer  2010-su      CS     225                Data Structures   
68990  2010  Summer  2010-su      CS     373          Theory of Computation   
68991  2010  Summer  2010-su      CS     421  Progrmg Languages & Compilers   
68992  2010  Summer  2010-su      CS     598  Object-Oriented Prog & Design   

      Sched Type   A+    A   A-  ...   C  C-  D+   D  D-   F  W  \
886          LEC    0  308   11  ...   2   3   1   0   1   4  1   
887          LCD  117  142   34  ...   6   4   2   1   4  14  0   
888          LEC   20  147   54  ...  26  10   8   8   5  16  5   
889          OLC    0  775  104  ...  17   0   0  24   0  22  2   
890          OLC    0  358   35  ...  17  13   0  17   0  33  9   
...          ...  ...  ...  ...  ...  ..  ..  ..  ..  ..  .. ..   
68988        LBD    4    6    2  ...   0   1   0   0   0   0  0   
68989        LBD    1    5    1  ...   3   1   0   2   0   1  0   
68990        LEC    5    1    5  ...   0   2   0   2   0   1  1   
68991        LCD    2    5    5  ...   1   4   0   4   0   0  1   
68992        ONL    3   22    0  ...   0   0   0   0   0   0  0   

        Primary Instructor  Students  A_Grades  
886      Fleck, Margaret M       347       319  
887      Beckman, Mattox A       372       293  
888          Zilles, Craig       422       221  
889    Challen, Geoffrey W      1062       879  
890         Nowak, Michael       578       393  
...                    ...       ...       ...  
68988    Gambill, Thomas N        21        12  
68989        Earls, John C        21         7  
68990         Kumar, Viraj        26        11  
68991       Hafiz, Munawar        30        12  
68992     Johnson, Ralph E        30        25  

[2214 rows x 24 columns]

### 🔬 Test Case Checkpoint 🔬

In [9]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df_STAT' in vars() ), "Make certain to name the STAT courses df_STAT."
assert( 'df_CS' in vars() ), "Make certain to name the CS courses df_CS."
assert( len(df_STAT[df_STAT.Subject != "STAT"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."
assert( len(df_CS[df_CS.Subject != "CS"] ) == 0 ), "It looks like you did not subset df_CS to only CS courses."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Puzzle 1.3: Comparing Overall Percentages

With our two new DataFrames of STAT and CS courses, use the following code cell to determine the **percentage** of A grades recieved in STAT and CS courses, storing them as variables `stat_a` and `cs_a` respectively. 

Print statements have been provided to show the values you calculate. 

As you work, remember:
- To find the **% of As**, divide the total number of `A_Grades` by the total number of `Students` ($\space \frac{A \space Grades}{Students} \space$).
- The `A_Grades` column you made earlier, and the `Students` column containing the total number of students in each course.
- The syntax `sum(df['column_name'])` can be used to add up the values of all rows in a particular column of a DataFrame.
- Your % should be a decimal between 0 and 1.

In [10]:
stat_a = sum(df_STAT["A_Grades"]) / sum(df_STAT["Students"])
print(f'Overall % of As in STAT is: {stat_a}')

cs_a = sum(df_CS["A_Grades"]) / sum(df_CS["Students"])
print(f'Overall % of As in CS is: {cs_a}')

Overall % of As in STAT is: 0.6151810949676277
Overall % of As in CS is: 0.6228580843585237


### 🔬 Test Case Checkpoint 🔬

In [11]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
import math
assert(math.isclose(stat_a, 0.6151810949676277)), "The overall percentage of A grades received in STAT courses does not appear to have been correctly calculated."
assert(math.isclose(cs_a, 0.6228580843585237)), "The overall percentage of A grades received in CS courses does not appear to have been correctly calculated."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Analysis: Good Comparison?
**Q: What conclusion can you take from the overall percentages found above when asking the original question:**
- *Is it significantly easier to get an A in STAT or CS courses at UIUC, or is it so close it would be hard to make general conclusions?*

We would say that the percentage is very close, and that CS has 1% more A's than STATS. Based on our data, we can say that it's inconclusive to determine which subject is easier to get an A in thus far.

**Q: Given what you've learned about experimental design, what are some reasons (specific to this dataset) you may not trust this conclusion? If you would trust it, explain why.**

There may be a concentration of students in one class that may be easier/harder to get an A in. This can be an example of Simpson's Paradox.

## An Extra Consideration

If you look at the `Year` column of our GPA dataset, you might notice that we have some old data in our dataset - all the way back to **2010**! This means we aren't really answering our question from the perspective of a student now. 

If we want to know if it is easier to **currently** get an A in a STAT or CS course, we should control for the **date** of the data by looking at more **recent years** specifically.

### Puzzle 1.4: More DataFrames

Using the code cells below, define four new DataFrames by selecting from rows of our previously created `df_CS` and `df_STAT`. 

- `df_stat_recent`: all `STAT` course data in *recent years* ($\space \geq 2021 \space$)
- `df_cs_recent`: all `CS` course data in *recent years* ($\space \geq 2021 \space$)
- `df_stat_old`: all other older `STAT` course data ($\space < 2021 \space$)
- `df_cs_old`: all other older `CS` course data ($\space < 2021 \space$)

Define *recent years* as **any year after and including 2021**, and *older years* as **any year before 2021**. 

Feel free to use conditionals OR the `.isin()` syntax you learned in the last lab.

In [12]:
df_stat_recent = df_STAT[df_STAT.Year >= 2021]

df_stat_recent

Year    Term YearTerm Subject  Number                   Course Title  \
2680   2023    Fall  2023-fa    STAT     100                     Statistics   
2681   2023    Fall  2023-fa    STAT     100                     Statistics   
2682   2023    Fall  2023-fa    STAT     100                     Statistics   
2683   2023    Fall  2023-fa    STAT     107         Data Science Discovery   
2684   2023    Fall  2023-fa    STAT     107         Data Science Discovery   
...     ...     ...      ...     ...     ...                            ...   
15623  2021  Winter  2021-wi    STAT     420      Statistical Modeling in R   
15624  2021  Winter  2021-wi    STAT     420  Methods of Applied Statistics   
15625  2021  Winter  2021-wi    STAT     420  Methods of Applied Statistics   
15626  2021  Winter  2021-wi    STAT     440    Statistical Data Management   
15627  2021  Winter  2021-wi    STAT     448         Advanced Data Analysis   

      Sched Type  A+    A   A-  ...   C  C-  D+  D  D-   F  W  \
2680         LCD  49  260  128  ...   9   2   2  0   3   8  0   
2681         LCD  44  177   77  ...  10   3   3  1   0   5  0   
2682         LCD   6   84   32  ...   5   2   2  0   0   0  0   
2683         LEC  56  285   59  ...  11   6   1  4   0  10  1   
2684         LEC  66  301   49  ...  11   6   2  2   0   3  1   
...          ...  ..  ...  ...  ...  ..  ..  .. ..  ..  .. ..   
15623        ONL  28  123   18  ...   1   0   0  0   0   1  0   
15624        ONL   5   14    5  ...   0   0   0  0   0   1  0   
15625        ONL  10   48   11  ...   1   0   0  0   0   2  1   
15626        ONL   0   29    0  ...   0   0   0  0   0   3  1   
15627        ONL   5    8    0  ...   1   0   0  0   0   1  0   

           Primary Instructor  Students  A_Grades  
2680          Rao, V.N. Vimal       608       437  
2681          Rao, V.N. Vimal       437       298  
2682             Clement, Kit       195       122  
2683        Flanagan, Karle A       524       400  
2684        Flanagan, Karle A       525       416  
...                       ...       ...       ...  
15623            Unger, David       185       169  
15624            Unger, David        25        24  
15625            Unger, David        86        69  
15626   Kinson, Christopher L        36        29  
15627  Bravo De Guenni, Lelys        21        13  

[275 rows x 24 columns]

In [13]:
df_cs_recent = df_CS[df_CS.Year >= 2021]

In [14]:
df_stat_old = df_STAT[df_STAT.Year < 2021]

In [15]:
df_cs_old = df_CS[df_CS.Year < 2021]

### 🔬 Test Case Checkpoint 🔬

In [16]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df_stat_recent' in vars() ), "Make certain to name the recent STAT courses df_stat_recent."
assert( 'df_cs_recent' in vars() ), "Make certain to name the recent CS courses df_cs_recent."
assert( 'df_stat_old' in vars() ), "Make certain to name the old STAT courses df_stat_old."
assert( 'df_cs_old' in vars() ), "Make certain to name the old CS courses df_cs_old."

assert( len(df_stat_recent[df_stat_recent.Year < 2021] ) == 0 ), "Make sure only years after and including 2021 are in the df_stat_recent DataFrame."
assert( len(df_cs_recent[df_cs_recent.Year < 2021] ) == 0 ), "Make sure only years after and including 2021 are in the df_cs_recent DataFrame."
assert( len(df_stat_old[df_stat_old.Year >= 2021] ) == 0 ), "Make sure only years before 2021 are in the df_stat_old DataFrame."
assert( len(df_cs_old[df_cs_old.Year >= 2021] ) == 0 ), "Make sure only years before 2021 are in the df_cs_old DataFrame."

assert( len(df[ df.index.isin(df_stat_recent.index) & df.index.isin(df_stat_old.index) ]) == 0 ), "Check for duplicate values in your df_stat_recent and df_stat_old DataFrames."
assert( len(df[ df.index.isin(df_cs_recent.index) & df.index.isin(df_cs_old.index) ]) == 0 ), "Check for duplicate values in your df_cs_recent and df_cs_old DataFrames."
assert( len(df_cs_old) + len(df_cs_recent) == len(df_CS) ), "You're excluding some rows from df_cs_old or df_cs_recent. Please double check your conditionals."
assert( len(df_stat_old) + len(df_stat_recent) == len(df_STAT) ), "You're excluding some rows from df_stat_old or df_stat_recent. Please double check your conditionals."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Puzzle 1.5: New Analysis
Now that we've got all the DataFrames setup with GPA data of the `CS` and `STAT` courses separated by recency (2021 or newer being 'recent'), we can do more in-depth analysis to investigate our question. 

In the following code cells, **calculate the percentages** described by the comment in the cell. Your answer should always be a **decimal between 0 and 1**. 

Just like Puzzle 1.3, remember: 
- To find the % of As, divide the total number of As by the total number of students. 
- The `A_Grades` column you made earlier, and the `Students` column containing the total number of students in each course.
- The syntax `sum(df['column_name'])` can be used to add up the values of all rows in a particular column of a DataFrame.

In [17]:
# Percentage of As received in CS in recent years
cs_recent_a = sum(df_cs_recent['A_Grades']) / sum(df_cs_recent['Students'])
print(f'Percentage of As received in CS in recent years: {cs_recent_a}')

Percentage of As received in CS in recent years: 0.7222567339230693


In [18]:
# Percentage of As received in STAT in recent years
stat_recent_a = sum(df_stat_recent['A_Grades']) / sum(df_stat_recent['Students'])
print(f'Percentage of As received in STAT in recent years: {stat_recent_a}')

Percentage of As received in STAT in recent years: 0.6628407460545194


In [19]:
# percentage of As received in CS in older years
cs_old_a = sum(df_cs_old['A_Grades']) / sum(df_cs_old['Students'])
print(f'Percentage of As received in CS in older years: {cs_old_a}')

Percentage of As received in CS in older years: 0.5716443898192247


In [20]:
# percentage of As received in STAT in older years
stat_old_a = sum(df_stat_old['A_Grades']) / sum(df_stat_old['Students'])
print(f'Percentage of As received in STAT in older years: {stat_old_a}')

Percentage of As received in STAT in older years: 0.5969278801123946


### 🔬 Test Case Checkpoint 🔬

In [21]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(math.isclose(cs_recent_a,  0.7222567339230693)), "The overall percentage of A grades received in CS courses recently does not appear to have been correctly calculated."
assert(math.isclose(stat_recent_a, 0.6628407460545194)), "The overall percentage of A grades received in STAT courses recently does not appear to have been correctly calculated."

assert(math.isclose(cs_old_a, 0.5716443898192247)), "The overall percentage of A grades received in CS courses in older years does not appear to have been correctly calculated."
assert(math.isclose(stat_old_a, 0.5969278801123946)), "The overall percentage of A grades received in STAT courses in older years does not appear to have been correctly calculated."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Observe the Results
**Run the following cell** to format all of your answers as a DataFrame.

Keep in mind that "Older" means data from courses **before 2021**, and "Recent" means courses held **during or after 2021**. 

In [22]:
pd.DataFrame([
  {'% of A (pre-2021)': cs_old_a, 'Recent % of A': cs_recent_a, 'Overall % of A': cs_a},
  {'% of A (pre-2021)': stat_old_a, 'Recent % of A': stat_recent_a, 'Overall % of A': stat_a}
], index=['CS', 'STAT'])

% of A (pre-2021)  Recent % of A  Overall % of A
CS             0.571644       0.722257        0.622858
STAT           0.596928       0.662841        0.615181

Notice that when observing the "Overall % of A" grades received:
- It appears that `STAT` and `CS` are **equally difficult** to get an `A_Grades` (62.28% vs. 61.52%).
- However, if you look at the sub-group of the courses held in years of **2021 and later**, we see that `CS` actually has a higher `A_Grades` rate! 

This is **Simpson's Paradox**: a pattern within a population can appear, disappear, or reverse when you look at subpopulations.

In more formal terms, Simpson's Paradox can cause you to observe a pattern reverse when you look at the overall group statistics versus statistics of groups post-stratification. In this case we are stratifying by time.


### Analysis: Reflecting on New Observations 

Now think about how would you now respond differently to the incoming student's question:
- *Is it easier to get an A in STAT or CS courses at UIUC?*

**Q: Which comparison of percentages do you trust more and why? Are there any other potential confounding variables when answering this question that could be investigated further? Respond with at least three full sentences.**

It seems that what's affecting our "Overall % of A's" value is the level of difficulty changing in recent years. Prior to 2021, STAT and CS had a similar distribution of A's, with CS being 2% smaller. In recent years, both percentages have increased dramatically: CS A's increasing by about 15% and STATS A's increasing by about 7%. Even though it is apparent that CS has a higher % of A's in recent years as compared to STATS, we see their Overall percentages being pretty similar. Our confounder can be the fact that Computer Science as a field has grown dramtically in recent years, and thus their department offers more introductory classes for beginners that may be easier to get an A in.

<hr style="color: #DD3403;">

# Part 2: Revisiting the Hello Dataset

Enough about GPA (for now). Two weeks ago, you created a series of questions that made up the **Hello Dataset** and completed the survey by answering all of the questions made by students in DISCOVERY. 

Now, we will load this dataset again and briefly answer a few questions with data about YOU! 

## Load the Hello Dataset

The "Hello Dataset" is available here:
```
https://waf.cs.illinois.edu/discovery/hello-fa24.csv
```

Use Python to load this dataset into a DataFrame called `df_hello`:

In [23]:
df_hello = pd.read_csv("https://waf.cs.illinois.edu/discovery/hello-fa24.csv")
df_hello

Name       Lab When flipping a coin, heads or tails?  \
0     Prof. Karle  AL1, AL2                                 Heads   
1      Prof. Wade  AL1, AL2                                 Tails   
2        Angelina       Y13                                 Tails   
3        TA Tyler       Y07                                 Tails   
4          Sneha        Y17                                 Tails   
...           ...       ...                                   ...   
1084        Peiyu       Y06                                 Heads   
1085    RuohanLiu       Y25                                 Tails   
1086       Morgan       Y03                                 Heads   
1087       Auggie       Y08                                 Heads   
1088       Rebeca       Y10                                 Tails   

     Is this your first year at UIUC?  \
0                      Not First Year   
1                      Not First Year   
2                  First Year at UIUC   
3                      Not First Year   
4                  First Year at UIUC   
...                               ...   
1084                   Not First Year   
1085                   Not First Year   
1086                   Not First Year   
1087                   Not First Year   
1088                   Not First Year   

     Is your favorite ice cream flavor chocolate or vanilla?  \
0                                             Chocolate        
1                                               Vanilla        
2                                               Vanilla        
3                                             Chocolate        
4                                               Vanilla        
...                                                 ...        
1084                                            Vanilla        
1085                                          Chocolate        
1086                                            Vanilla        
1087                                          Chocolate        
1088                                          Chocolate        

     Do you like morning or night?  Do you prefer even or odd numbers?  \
0                             Night                               Even   
1                             Night                                Odd   
2                           Morning                               Even   
3                             Night                               Even   
4                             Night                               Even   
...                             ...                                ...   
1084                          Night                               Even   
1085                          Night                                Odd   
1086                          Night                                Odd   
1087                          Night                                Odd   
1088                        Morning                               Even   

     Do you have a pet? Do you prefer summer or winter?  \
0                    No                          Summer   
1                   Yes                          Winter   
2                   Yes                          Summer   
3                    No                          Winter   
4                    No                          Summer   
...                 ...                             ...   
1084                 No                          Summer   
1085                Yes                          Winter   
1086                 No                          Summer   
1087                Yes                          Summer   
1088                Yes                          Summer   

     Are you a business major?  ...  \
0                           No  ...   
1                           No  ...   
2                           No  ...   
3                           No  ...   
4                          Yes  ...   
...                        ...  ...   
1084                        No  ...   
1085        

### 🔬 Test Case Checkpoint 🔬

In [24]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(len(df_hello) >= 1089), "This is not the Hello Dataset you're looking for. Check the URL."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Number of Classes vs. Ideal Party Size

There is a myth that you can either study or party in college, but not both. Let's explore if this is true or not in our data!

In the Hello Dataset, there are two questions with numeric responses that can help us explore this:
- How many credit hours are you taking?
- How many people is the ideal number of people at a party?

### Puzzle 2.1: Observation Subsets

In this situation, let's define three levels of course load based on credit hours:
- You have **light course load** if you have 14 or fewer credit hours,
- You have a **medium course load** if you have 15 or 16 credit hours,
- Otherwise, you have a **heavy course load** (with 17 or more credit hours).

From this, **create three DataFrames** that contain subsets of the Hello Dataset: 
- `df_light_course_load`: including everyone who has a "light course load" ($\space \leq 14 \space$)
- `df_medium_course_load`: including everyone who has a "medium course load" ($\space >14$ and $<17$)
- `df_heavy_course_load`: including everyone who has a "heavy course load" ($\space  \geq 17 \space$)

The credit hour question has been provided to you as a **string** stored in the variable `credit_hour_question` for ease of DataFrame column access.

In [35]:
credit_hour_question = "How many credit hours are you taking?"
print(df_hello[credit_hour_question])
df_light_course_load = df_hello[df_hello[credit_hour_question] <= 14.0]
df_light_course_load


0        0.0
1        NaN
2       17.0
3       12.0
4       16.0
        ... 
1084    12.0
1085    17.0
1086    18.0
1087    17.0
1088    13.0
Name: How many credit hours are you taking?, Length: 1089, dtype: float64


Name       Lab When flipping a coin, heads or tails?  \
0     Prof. Karle  AL1, AL2                                 Heads   
3        TA Tyler       Y07                                 Tails   
6           Mark        Y04                                 Heads   
13       Kristina       Y33                                 Tails   
36         Aditya       Y27                                 Tails   
...           ...       ...                                   ...   
1061    TA Kunlun  Y12, Y25                                 Heads   
1063        Shlok       Y04                                 Heads   
1076       Ishaan       Y19                                 Heads   
1084        Peiyu       Y06                                 Heads   
1088       Rebeca       Y10                                 Tails   

     Is this your first year at UIUC?  \
0                      Not First Year   
3                      Not First Year   
6                      Not First Year   
13                     Not First Year   
36                 First Year at UIUC   
...                               ...   
1061               First Year at UIUC   
1063               First Year at UIUC   
1076               First Year at UIUC   
1084                   Not First Year   
1088                   Not First Year   

     Is your favorite ice cream flavor chocolate or vanilla?  \
0                                             Chocolate        
3                                             Chocolate        
6                                             Chocolate        
13                                              Vanilla        
36                                            Chocolate        
...                                                 ...        
1061                                          Chocolate        
1063                                          Chocolate        
1076                                            Vanilla        
1084                                            Vanilla        
1088                                          Chocolate        

     Do you like morning or night?  Do you prefer even or odd numbers?  \
0                             Night                               Even   
3                             Night                               Even   
6                           Morning                               Even   
13                            Night                               Even   
36                          Morning                               Even   
...                             ...                                ...   
1061                        Morning                               Even   
1063                          Night                               Even   
1076                        Morning                               Even   
1084                          Night                               Even   
1088                        Morning                               Even   

     Do you have a pet? Do you prefer summer or winter?  \
0                    No                          Summer   
3                    No                          Winter   
6                    No                          Winter   
13                   No                          Winter   
36                   No                          Summer   
...                 ...                             ...   
1061                Yes                          Summer   
1063                 No                          Summer   
1076                Yes                          Winter   
1084                 No                          Summer   
1088                Yes                          Summer   

     Are you a business major?  ...  \
0                           No  ...   
3                           No  ...   
6                           No  ...   
13                          No  ...   
36                          No  ...   
...                        ...  ...   
1061                        No  ...   
1063        

In [30]:
df_medium_course_load = df_hello[(df_hello[credit_hour_question] > 14.0) & (df_hello[credit_hour_question] < 17.0)]
df_medium_course_load

Name  Lab When flipping a coin, heads or tails?  \
4        Sneha   Y17                                 Tails   
5          Gage  Y10                                 Heads   
7         Jaden  Y34                                 Heads   
9        Amanda  Y23                                 Heads   
10    Stephanie  Y36                                 Heads   
...         ...  ...                                   ...   
1077      Jenny  Y32                                 Heads   
1078      Aidan  Y08                                 Heads   
1081       Kyle  Y33                                 Tails   
1082      Isaac  Y04                                 Tails   
1083      Chloe  Y20                                 Tails   

     Is this your first year at UIUC?  \
4                  First Year at UIUC   
5                  First Year at UIUC   
7                  First Year at UIUC   
9                  First Year at UIUC   
10                 First Year at UIUC   
...                               ...   
1077                   Not First Year   
1078                   Not First Year   
1081               First Year at UIUC   
1082                   Not First Year   
1083               First Year at UIUC   

     Is your favorite ice cream flavor chocolate or vanilla?  \
4                                               Vanilla        
5                                               Vanilla        
7                                             Chocolate        
9                                               Vanilla        
10                                            Chocolate        
...                                                 ...        
1077                                          Chocolate        
1078                                          Chocolate        
1081                                          Chocolate        
1082                                          Chocolate        
1083                                            Vanilla        

     Do you like morning or night?  Do you prefer even or odd numbers?  \
4                             Night                               Even   
5                             Night                               Even   
7                           Morning                               Even   
9                             Night                               Even   
10                            Night                               Even   
...                             ...                                ...   
1077                          Night                               Even   
1078                          Night                               Even   
1081                          Night                                Odd   
1082                        Morning                                Odd   
1083                          Night                               Even   

     Do you have a pet? Do you prefer summer or winter?  \
4                    No                          Summer   
5                   Yes                          Winter   
7                    No                          Summer   
9                   Yes                          Summer   
10                   No                          Summer   
...                 ...                             ...   
1077                Yes                          Summer   
1078                Yes                          Summer   
1081                Yes                          Summer   
1082                 No                          Summer   
1083                Yes                          Summer   

     Are you a business major?  ...  \
4                          Yes  ...   
5                           No  ...   
7                           No  ...   
9                          Yes  ...   
10                          No  ...   
...                        ...  ...   
1077                        No  ...   
1078                        No  ...   
1081                        No  ...   
1082            

In [31]:
df_heavy_course_load = df_hello[df_hello[credit_hour_question] >= 17.0]
df_heavy_course_load

Name  Lab When flipping a coin, heads or tails?  \
2      Angelina  Y13                                 Tails   
8          Eman  Y11                                 Heads   
11        Najam  Y20                                 Tails   
19        Brady  Y16                                 Heads   
23      Krystal  Y02                                 Heads   
...         ...  ...                                   ...   
1079      Joyce  Y21                                 Heads   
1080     Yunfan  Y25                                 Heads   
1085  RuohanLiu  Y25                                 Tails   
1086     Morgan  Y03                                 Heads   
1087     Auggie  Y08                                 Heads   

     Is this your first year at UIUC?  \
2                  First Year at UIUC   
8                  First Year at UIUC   
11                 First Year at UIUC   
19                 First Year at UIUC   
23                 First Year at UIUC   
...                               ...   
1079               First Year at UIUC   
1080                   Not First Year   
1085                   Not First Year   
1086                   Not First Year   
1087                   Not First Year   

     Is your favorite ice cream flavor chocolate or vanilla?  \
2                                               Vanilla        
8                                             Chocolate        
11                                            Chocolate        
19                                              Vanilla        
23                                              Vanilla        
...                                                 ...        
1079                                          Chocolate        
1080                                          Chocolate        
1085                                          Chocolate        
1086                                            Vanilla        
1087                                          Chocolate        

     Do you like morning or night?  Do you prefer even or odd numbers?  \
2                           Morning                               Even   
8                           Morning                               Even   
11                            Night                                Odd   
19                          Morning                                Odd   
23                            Night                                Odd   
...                             ...                                ...   
1079                          Night                               Even   
1080                        Morning                               Even   
1085                          Night                                Odd   
1086                          Night                                Odd   
1087                          Night                                Odd   

     Do you have a pet? Do you prefer summer or winter?  \
2                   Yes                          Summer   
8                    No                          Summer   
11                   No                          Winter   
19                   No                          Summer   
23                   No                          Summer   
...                 ...                             ...   
1079                 No                          Winter   
1080                Yes                          Summer   
1085                Yes                          Winter   
1086                 No                          Summer   
1087                Yes                          Summer   

     Are you a business major?  ...  \
2                           No  ...   
8                           No  ...   
11                          No  ...   
19                         Yes  ...   
23                          No  ...   
...                        ...  ...   
1079                       Yes  ...   
1080                        No  ...   
1085                        No  ...   
1086            

### 🔬 Test Case Checkpoint 🔬

In [36]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(len(df_light_course_load) == 171), "Check your conditionals for `df_light_course_load`."
assert(len(df_medium_course_load) == 470), "Check your conditionals for `df_medium_course_load`."
assert(len(df_heavy_course_load) == 447), "Check your conditionals for `df_heavy_course_load`."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Puzzle 2.2: Overall Ideal Party Size

Find the average ideal party size among **everyone** in the Hello Dataset:

- The `df['column name'].mean()` function returns the mean of all values in the specified column of `df`.
- The ideal party size question has been provided to you as a **string** stored in the variable `party_size_question` for ease of DataFrame column access.

In [38]:
party_size_question = 'How many people is the ideal number of people at a party?'

overall_avg_ideal_party_size = df_hello[party_size_question].mean()
overall_avg_ideal_party_size

np.float64(33.91528545119705)

### Puzzle 2.3: Ideal Party Size Stratified by Course Load

Now find the average **ideal party size** of each group (`df_light_course_load`, `df_medium_course_load`, and `df_heavy_course_load`):

In [42]:

light_course_load_avg_ideal_party_size = df_light_course_load[party_size_question].mean()
light_course_load_avg_ideal_party_size

np.float64(32.98235294117647)

In [43]:
medium_course_load_avg_ideal_party_size = df_medium_course_load[party_size_question].mean()
medium_course_load_avg_ideal_party_size

np.float64(31.963675213675213)

In [44]:
heavy_course_load_avg_ideal_party_size = df_heavy_course_load[party_size_question].mean()
heavy_course_load_avg_ideal_party_size

np.float64(36.380313199105146)

### 🔬 Test Case Checkpoint 🔬

In [45]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(math.isclose(overall_avg_ideal_party_size, 33.91528545119705)), "You calculated your overall ideal party size mean incorrectly."
assert(math.isclose(light_course_load_avg_ideal_party_size, 32.98235294117647)), "You calculated your `df_light_course_load` ideal party size mean incorrectly."
assert(math.isclose(medium_course_load_avg_ideal_party_size, 31.963675213675213)), "You calculated your `df_medium_course_load` ideal party size mean incorrectly."
assert(math.isclose(heavy_course_load_avg_ideal_party_size, 36.380313199105146)), "You calculated your `df_heavy_course_load` ideal party size mean incorrectly."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Analysis: Course Load vs. Party Size

Run the following cell to see a summary of everything you found:

In [46]:
pd.DataFrame([
  {"Statistic": "=== Overall Results ===", "Value": ""},
  {"Statistic": "Overall Ideal Party Size", "Value": str(round(overall_avg_ideal_party_size, 2)) + " people"},
  {"Statistic": "=== Stratified Results ===", "Value": ""},
  {"Statistic": "Party Size (Light Course Load)", "Value": str(round(light_course_load_avg_ideal_party_size, 2)) + " people"},
  {"Statistic": "Party Size (Medium Course Load)", "Value": str(round(medium_course_load_avg_ideal_party_size, 2)) + " people"},
  {"Statistic": "Party Size (Heavy Course Load)", "Value": str(round(heavy_course_load_avg_ideal_party_size, 2)) + " people"},
])

Statistic         Value
0          === Overall Results ===              
1         Overall Ideal Party Size  33.92 people
2       === Stratified Results ===              
3   Party Size (Light Course Load)  32.98 people
4  Party Size (Medium Course Load)  31.96 people
5   Party Size (Heavy Course Load)  36.38 people

**Q: Is there any relationship between the number of credit hours and the ideal party size?  Is this a case of Simpson's Paradox?**

Explain why there is or is not Simpson's Paradox in at least one sentence.  Make sure to be specific on why you found or did not find Simpson's Paradox.

It seems that those who have a heavier course load find that the ideal party size is 36 people, while those taking less than 17 credit hours think that the ideal party size is around 32 people. It may be a case of Simpson's Paradox, as there could be other factors that tie in. For example, the types of courses that people with a heavy course load are taking. I would want to consider factors such as course levels and such, this will tell us if such rigorous students find the ideal party size to be larger or not.

<hr style="color: #DD3403;">

# Submission

You're almost done!  All you need to do is to commit your lab to GitHub:

1.  ⚠️ **Make certain to save your work.** ⚠️ To do this, go to **File => Save All**

2.  After you have saved, exit this notebook and follow the Canvas instructions to commit this lab to your Git repository!

3. Your TA will grade your submission and provide you feedback after the lab is due. :)